### PR TITLE
Button: Disabled buttons now receive focus

### PR DESCRIPTION
**Pull request by dzearing**
#### Pull request checklist

- [X] Addresses an existing issue: #912
- [X] Include a change request file if publishing 
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [ ] Documentation update

#### Description of changes

Buttons previously relied on the `disabled` state to prevent focus from being received. The normal browser behavior is to prohibit focus from being set to disabled buttons, even when `tabindex` is set to 0. However this conflicts with what's said in this spec:

https://www.w3.org/TR/wai-aria-practices-1.1/#menu

So to be good citizens here, the proposal and fix here is to make all disabled buttons focusable by avoiding the disabled attribute completely.

Disabled buttons will have the correct styling, and will not react to clicks.

While I'm not completely resolving issue #912, this is a pre-requisite for fixing that. The next fix is to consolidate all button usage to use `BaseButton` and provide variant css.

### CODE REVIEW

**Comment by joschect** [comment]
[Lines 47-49](https://github.com/dzearing/office-ui-fabric-react/blob/6414f9344ab5f47fefd09950fc551dd88294524c/packages/office-ui-fabric-react/src/components/Button/BaseButton.tsx#L47-L49) seem to suggest that it was intentional to allow focus.

File: packages/office-ui-fabric-react/src/components/Button/PrimaryButton/PrimaryButton.scss [hunk 1]
```diff
@@ -17,35 +17,27 @@
...
-
-.ms-Fabric.is-focusVisible .ms-Button--primary {
-  &:focus {
-    color: $ms-color-white;
-
-    &:before {
-      border-color: $ms-color-themeLighter;
-    }
+  &.is-disabled {
+    @include focus-border();
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/PrimaryButton/PrimaryButton.scss]: Shouldn't focus-border be included regardless of disabled state?

File: packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss [hunk 2]
```diff
@@ -58,63 +52,3 @@
...
-
-    .ms-Button-label {
-      color: $ms-color-themePrimary;
-    }
-  }
-
-  &:disabled,
-  &.is-disabled {
-    .ms-Button-icon {
-      color: $ms-color-neutralTertiaryAlt;
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss]: Why did you remove the disabled stuff from here? It really seems to cut down on code duplication. Now in every other *Button.scss you have to have &.is-disabled with `color: $ms-color-neutralTertiaryAlt;` Additionally you'll have to change it in all those places if it every changes and it makes it harder for me to subclass easily, I'll have to rewrite the css every time.

**Comment by joschect** [comment]
One other thing to consider, you can no longer check button.disabled since you aren't adding that to the element.

**Comment by mdahamiwal** [comment]
Does that mean that disabled buttons are always in tab order? https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/#focus_tabindex, clearly specify that only in case of grouped navigation widgets (eg. a menubar), we should allow arrow key navigation to disabled elements to provide discovery for ATs. This also applies if menu item are links, how are we planning to solve it for links?

**Comment by dzearing** [comment]
@mdahamiwal Thanks for this mention. I think it is unclear when buttons are used in a "navigation widget" or "menu bar". For example, is a dialog's OK/Cancel buttons part of navigation? I might argue yes; they dismiss the dialog and optionally perform some action, or go to the next step in the wizard.

So then, is there a scenario where a disabled button shouldn't be focusable? Does it cause user confusion?

I believe the spec is rather ambiguous here, and that we should opt on the side of usability than for trying to chase an ambiguous definition. The only other option i can think of is to make the focusable behavior configurable, but in my opinion that creates bugs since its left up to interpretation. Do it always or don't do it. I can understand the argument for doing it always but I can't defend never doing it.

**Comment by mdahamiwal** [comment]
Okay, I see your point but these are the aria standards as per my understanding and what we follow:
1. There are a few aria design patterns like menubar where this is essential for the reason that when ATs encounter these widget they announce the number of total number of items in it without actually announcing the name/label of individual items. Now, if these items are not focusable (using ↔️ ), it will be extremely difficult for the impaired users to know the actions he can perform.

2. Outside of such controls, any actionable elements or controls (a, button, any other widget) should not be part of tab order if they are disabled (until unless their design pattern is explicit about it). For dialog scenario that you mentioned, we normally don't disable `cancel` button so it will always be there. The CTA(OK/Submit/Save) button might be disabled due to pending selections, information in the dialog but since the user generally follow the form flow, he will have an idea why CTA buttons are not coming up. 

Hope that clarifies it.

**Comment by mdahamiwal** [comment]
Now for this particular change, if commandbar is composed of `button` we need to make this behavior optional and set it when using menubars since adding a `disabled` attribute will not work. Generally to avoid this hack on native elements, we use `<li>`s as `menu-item` roles and its associated aria-* attributes.

**Comment by dzearing** [comment]
Hmm. So it sounds like you're saying, make it optional and default it to false? And only use it in the command bar case, the contextual menu case, and maybe the nav bar case?

Keep in mind after this checkin, I'd like us to fix command bar to use this so we have a single button implementation.

File: packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss (hunk 2 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss]: uh... this is the hero button styling, which can be removed.

**Comment by mdahamiwal** [comment]
> So it sounds like you're saying, make it optional and default it to false? And only use it in the command bar case, the contextual menu case, and maybe the nav bar case?

That's right, the behavior should be guided by aria design patterns. A commandbar, navbar or context menu should follow menubar design pattern. Currently NavBar is authored as a List which IMO is not  correct since a NavBar has actions (navigations), sub-menus in it. It align more towards a menubar design.

> While I can do that, I still feel like this leaves it up to interpretation.

If we talk about user experience, the users perceive these as menu items (in case of menuBar) and not buttons and they are well aware of MenuBar design patterns, so nothing is left to interpretation.

> But if we don't wholesale make a consistent behavior, we end up with unpredictability.

There might be a implementation level unpredictability (since we are using button as menu items) and it would require good test coverage around it. Like suggested earlier, we can create a separate component `MenuItem` and avoid adding more code in `Button` or any other stable components.

File: packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss [hunk 3]
```diff
@@ -31,18 +31,12 @@
     border-color: $ms-color-contrastWhiteSelected;
   }
 
-  &:disabled,
-  &.disabled {
+  &.is-disabled {
```
Comment (mdahamiwal) [inline packages/office-ui-fabric-react/src/components/Button/ButtonCore/ButtonCore.scss]: Can we use attribute selectors eg. `aria-disabled=true` to provide styles wherever possible? It provides multiple benefits:
1. Provides symantic to our CSS - CSS is more readable and meaningful.
2. Less chances to accessibility regression - any regression in accessibility due to future changes can be caught upfront.

**Comment by micahgodbolt** [comment]
@dzearing I know you've got a lot going, but there are a ton of conflicts here. Do you have time to tackle them, or want me to dive in?

**Comment by dzearing** [comment]
Closing will revisit.

**Pull request closed without merging**